### PR TITLE
macOS: Enable hardened runtime; notarize during build

### PR DESCRIPTION
**Pull request by jmcphers**
This change adds [Apple notarization](https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution) to the RStudio Desktop build for MacOS. Apple has indicated that this will soon be a requirement for all applications not distributed through the App Store. This is an act in three parts:

1. During code signing, we pass a flag opting in to the hardened runtime (only hardened apps can be notarized). This restricts the app's behavior in various ways for security reasons. Not surprisingly, RStudio does not run with the default hardened runtime configuration.
2. We add a configuration that gives us [entitlements](https://developer.apple.com/documentation/security/hardened_runtime_entitlements), which exclude us from some hardened runtime restrictions. I've pared these down by trial and error to the minimal set for basic operation; we may need to add more later.
3. In the build lab, prior to uploading the build to S3, we send it to Apple for notarization. Apple runs a public service that currently takes ~5 minutes to notarize RStudio. Once complete, we staple the notarization ticket to our binary so that it can be validated offline (without this step macOS has to contact a web service to see if there is a record of the notarization).

Note that this reintroduces https://github.com/rstudio/rstudio/issues/2313; with the hardened runtime on, the R session doesn't start when it fails to find the dummy R library in `@executable_path`. We will need to find another way to fix that issue, possibly by pointing to a version of R that doesn't exist. 

Fixes https://github.com/rstudio/rstudio/issues/4597.

### CODE REVIEW

File: src/cpp/session/CMakeLists.txt [hunk 1]
```diff
@@ -625,13 +625,6 @@ if(APPLE)
          add_custom_command (TARGET rsession POST_BUILD
             COMMAND install_name_tool -change ${LIB_DIR}/${LIB_FILE} @executable_path/../Frameworks/${LIB_FILE} ${CMAKE_CURRENT_BINARY_DIR}/rsession)
       endforeach()
-
-      # also rewrite where rsession looks for libR.dylib
-      # (so we don't accidentally load the wrong version of R at launch)
-      get_filename_component(LIBR_HOME_REALPATH "${LIBR_HOME}" REALPATH)
-      add_custom_command(TARGET rsession POST_BUILD
-         COMMAND install_name_tool -change "${LIBR_HOME_REALPATH}/lib/libR.dylib" "@executable_path/libR.dylib" "${CMAKE_CURRENT_BINARY_DIR}/rsession")
```
Comment (kevinushey) [inline src/cpp/session/CMakeLists.txt]: We could consider whether rewriting this to be `/Library/Frameworks/R.framework/Versions/Current/Resources/lib/libR.dylib`?

Comment (kevinushey) [reply src/cpp/session/CMakeLists.txt]: (assuming we're allowed to have library link paths outside of the application bundle)

Comment (jmcphers) [reply src/cpp/session/CMakeLists.txt]: Good idea, will give that a shot.

Comment (jmcphers) [reply src/cpp/session/CMakeLists.txt]: Update: works pretty well!

Comment (kevinushey) [reply src/cpp/session/CMakeLists.txt]: 🎉 🎉 🎉 🎉 🎉

**Comment by anatoly-os** [comment]
Hello! The changes you've made to make dynamic libraries work in runtime just saved pending [MuseScore](https://github.com/musescore/MuseScore) release :) Thank you very much from MuseScore's community!

**Comment by jmcphers** [comment]
@anatoly-os I'm delighted to hear that, as I'm a pianist myself and use MuseScore several times a week!

**Pull request merged**